### ISSUE TITLE
[BUG] LabelEncoder returns not fitted error in a nested pipeline in predict

### ISSUE THREAD

**Issue by Abhay-Lejith** [sktime/sktime#6463]
**Describe the bug**

Important: this is after incorporating the fix for scitype issue in #6458. This error is not related to it but for reproducing this error, code changed in that pr is necessary.
LabelEncoder returns a not fitted error when used in a nested pipeline after calling predict.
**To Reproduce**

```python
import numpy
import pandas
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sktime.forecasting.arima import ARIMA
from sktime.transformations.series.subset import ColumnSelect

different_series = ["A", "B"]
different_dates = pandas.date_range(start="2000-01-01", end="2000-03-31", freq="D")

number_of_series = len(different_series)
number_of_timesteps = len(different_dates)

random_generator = numpy.random.default_rng(seed=0)

different_values = [1, 2, 3]

sample_data = {
    "series": numpy.repeat(different_series, number_of_timesteps),
    "dates": numpy.tile(different_dates, number_of_series),
    "P": random_generator.standard_normal(size=number_of_series * number_of_timesteps),
    "Q": random_generator.choice(
        different_values, size=number_of_series * number_of_timesteps, replace=True
    ),
    "R": random_generator.standard_exponential(size=number_of_series * number_of_timesteps),
}

sample_dataset = pandas.DataFrame(data=sample_data)
sample_dataset = sample_dataset.set_index(["series", "dates"])

pipeline = (ColumnSelect(columns=["Q"]) * LabelEncoder()) ** (StandardScaler() * ARIMA())
pipeline.fit(sample_dataset[["P"]], X=sample_dataset[["Q", "R"]])
z=pandas.DataFrame({"Q":[1,2,3], "R":[2.3,1.2,1.8]})
pipeline.predict(fh=[1,2,3],X=z)
```

**Expected behavior**

Should run without error.
**Additional context**

Though the above pipeline doesnt work, the following does:
```python
pipeline1 = (ColumnSelect(columns=["Q"]) * LabelEncoder())
pipeline1.fit(sample_dataset[["Q", "R"]])
pipeline1.transform(sample_dataset[["Q", "R"]])
```
**Versions**
<details>
System:
    python: 3.10.12 (main, Jul  5 2023, 18:54:27) [GCC 11.2.0]
   machine: Linux-6.5.0-35-generic-x86_64-with-glibc2.35

Python dependencies:
          pip: 24.0
       sktime: 0.29.0
      sklearn: 1.3.2
       skbase: 0.5.2
        numpy: 1.24.3
        scipy: 1.11.3
       pandas: 2.0.2
   matplotlib: 3.7.2
       joblib: 1.3.2
        numba: 0.58.1
  statsmodels: 0.14.0
     pmdarima: 2.0.3
statsforecast: 1.5.0
      tsfresh: 0.20.1
      tslearn: 0.5.3.2
        torch: 2.2.1+cu121
   tensorflow: 2.13.0
tensorflow_probability: None

</details>

**Comment by fkiraly** [comment]
@Abhay-Lejith, thanks for reporting. Can you specify which branch this is on? Is it the current state of #6458? Or is it on `main`?

**Comment by Abhay-Lejith** [comment]
It is on the current state of #6458 .

**Comment by fkiraly** [comment]
Then it's not strictly speaking a bug, the tracker is for reporting issues on `main`.

If about the state of a PR, the PR is the best place to discuss, unless you think it's going to be a long, separate discussion.

I'll rename the issue to avoid confusion - we can discuss here how to address things.

**Comment by fkiraly** [comment]
Hm, ok, correcting myself - it *could* actually be a bug, that is only triggered given current state of #6458.

Can you explain though why you think it is a bug and not an issue with the one object that is being changed, in #6458?

**Comment by fkiraly** [comment]
I quickly checked, it looks like the `TabularToSeriesAdaptor` does not fit it. Seems like an issue with the focus of #6458, so you may want to debug it in the PR.

**Comment by Abhay-Lejith** [comment]
#6458 fixes the scitype error reported in issue #5867 which prevented the encoder from working at all. This error(of not getting fit) occurs after solving the scitype error and only in nested pipelines. So I just thought since its a different error, I'd open a new issue. 

> so you may want to debug it in the PR.

Sure, I'll continue debugging and make commits in the same pr then.

**Comment by fkiraly** [comment]
I see - yes, *might* be a different bug then.

**Comment by Abhay-Lejith** [comment]
Not quite sure whats happening here, my observations so far:

Basically during predict `LabelEncoder` first checks if it is fitted by checking if it has a `classes_ ` attribute. Error is because this is not present.
Tried this in the `_fit` method in forecasting pipeline file i.e. `_pipeline.py`.
```python
if not skip_trafos:
    # transform X
    for step_idx, name, transformer in self._iter_transformers():
        t = transformer.clone()
        X = t.fit_transform(X=X, y=y)
        if hasattr(t, "transformer_"):   # modification to check if classes_ is present
            print(t.transformer_.classes_)   
        self.steps_[step_idx] = (name, t)
```
This throws an error which means the encoder ( `t.transformer_` here) does not have `classes_` attribute right after the fit_transform in the _fit method of pipeline itself.

But after trying the same print statement right after the transform method in the adaptor, it is present there.

When `t.fit_transform` is called , isn't the same instance being used in the adaptor?

**Comment by fkiraly** [comment]
> When `t.fit_transform` is called , isn't the same instance being used in the adaptor?

It should be - I checked the pipeline layer and it looked fine, but this is not conclusive.

The difficulty lies in how the classes are nested:

* `TabularToSeriesAdaptor` inside the pipeline (user does not see this but it is there)
* `LabelEncoder` inside the `TabularToSeriesAdaptor`

I would recommend you keep track of the `dir`-s of both when calling `fit` and at the start of `transform`, and first narrow it down to the level where the issue occurs, i.e., is it the pipeline or the adaptor?

**Comment by Abhay-Lejith** [comment]
That's the weird part, at the end of the transform method of the adaptor, the encoder has the `classes_` attribute indicating that it is fit. But immediately after this, when it returns to the pipeline, the encoder no longer has the `classes_` attribute.

**Comment by fkiraly** [comment]
in the above statement, can you kindly explain, which of the objects are you referring to precisely? There are three: the pipeline, the adaptor, the `LabelEncoder`. Which one is trying to access an attribute that it does not have, when? Which one does have the attribute, when?

**Comment by Abhay-Lejith** [comment]
I am referring to the `LabelEncoder` trying to access its `classes_` attribute.
In the code snippet above, `t` is the adaptor, `t.transformer_`  is the encoder. 
Also just mentioning again that the code snippet is in the `fit` method of `pipeline`.

Why this attribute is important is because it indicates whether the encoder has been fit or not.

This line 
```python
X=t.fit_transform(X=X, y=y)
````
 would call fit and transform methods of the label encoder eventually(i.e after going through base class). I tested whether at the end of the transform method(in adaptor where transform of encoder is called) classes_ is present and it was present. After this it should return to the next line i.e
````python
if hasattr(t, "transformer_"):   
            print(t.transformer_.classes_)  
```
which gave error indicating encoder didnt have classes_ attribute.
I hope it is clearer now.

**Comment by Abhay-Lejith** [comment]
@fkiraly should pr #6458 be added to the project board? Though all tests pass and it fixes #5867 , since I will be making commits for this issue in the same pr as well, it is still 'under progress' .

**Comment by Abhay-Lejith** [comment]
> Sure . I am referring to the `LabelEncoder` trying to access its `classes_` attribute.
> In the code snippet above, `t` is the adaptor, `t.transformer_`  is the encoder. 
> Also just mentioning again that the code snippet is in the `fit` method of `pipeline`.
> 
> Why this `classes_` attribute is important is because it indicates whether the encoder has been fit or not.
> 
> This line 
> ```python
> X=t.fit_transform(X=X, y=y)
> ```
>  would call fit and transform methods of the label encoder eventually(i.e after going through base class). I tested whether at the end of the transform method(in adaptor where transform of encoder is called) classes_ is present and it was present. After this it should return to the next line i.e
> ```python
> if hasattr(t, "transformer_"):   
>             print(t.transformer_.classes_)  
> ```
> which gave error indicating encoder didnt have classes_ attribute anymore. Although it just did at the end of the transform method.
> I hope it is clearer now.
> 
> 

@fkiraly Any thoughts on this now?

**Comment by fkiraly** [comment]
Apologies, the explanation is still a bit nebulous to me, I need this information:

> can you kindly explain, which of the objects are you referring to precisely? There are three: the pipeline, the adaptor, the LabelEncoder. Which one is trying to access an attribute that it does not have, when? Which one does have the attribute, when?

So, which object runs which code, and in which object is which other object missing which attribute, in which state?

**Comment by Abhay-Lejith** [comment]
I think it might be easier for me to explain and for you to understand this via a meet on discord perhaps where I can screenshare the code and explain it? Is setting up a short meet (5-10 mins) tomorrow i.e. Sunday possible? Otherwise we can do it during the standup/mentoring session on Monday.

**Comment by fkiraly** [comment]
Yes, we can do that - could you try writing a structured summary first? That would still be helpful even if we set up a sync meet.

**Comment by Abhay-Lejith** [comment]
Ok. Let me summarize.
No error when pipeline `fit` is called.
Error occurs when pipeline `predict` is called.
More specifically, when the `transform` of `LabelEncoder` is called during pipeline predict.
This error happens because `LabelEncoder` cannot access its `classes_` attribute. This attribute is present only if it has been fit. So it is basically a test to check if encoder has already been fitted.

So I try to find at what stage in the pipeline fit/predict, the encoder does not possess this attribute.
I find that this happens in pipeline `fit` itself.
More specifically, at this point in the code of pipeline fit.
```python
if not skip_trafos:
    # transform X
    for step_idx, name, transformer in self._iter_transformers():
        t = transformer.clone()
        X = t.fit_transform(X=X, y=y)
        if hasattr(t, "transformer_"):   # I added these two lines to check for the attribute.
            print(t.transformer_.classes_)   # modification to check if classes_ is present 
        self.steps_[step_idx] = (name, t)
```
Here, `t` is the `adaptor`, `t.transformer_` is the `encoder`.
Now when I run the code, it fails, indicating that the `classes_` attribute is not present.

So , next I try to see where in the previous line i.e  `X = t.fit_transform(X=X, y=y)`,  is the `classes_` attribute missing. 
Once again, `t` is the adaptor here.

In the adaptors `transform` method, the transform method of the encoder is called. I check if `classes_` attribute is present after this transform method. I find that it is present.

So problem here: When X = t.fit_transform(X=X, y=y) is called, the classes_ attribute is present at the end of the adaptors transform method. But once it returns to this pipeline code, it is no longer there.

**Comment by fkiraly** [comment]
Could you check after `fit`, and just before the error, the identity of:

* the pipeline
* the adaptor, inside the pipeline
* the `transformer_` attribute
* the `transformer_`'s `classes_` attribute?

Are these the same objects? I.e., identical (as via `is` or - probably easier to check via - `id`), or just instances of the same class but non-identical?

Also, could you check the adaptor's `fit_is_empty` tag via `get_tags()` just before the error?

**Comment by Abhay-Lejith** [comment]
Never mind, the issue was because the `X` I passed to `pipeline.predict` was not in the same panel format as the `X` passed to `fit`. This caused the regular transform to be called instead of the vectorized version. This caused the error.

Everything works now after passing proper X to predict. No changes to code needed.

**Comment by fkiraly** [comment]
ah, interesting. Perhaps it is worth ensuring that an informative message is raised. I think this has caused enough confusion that we could consider that the actual problem.

That is, if scitypes in fit/predict disagree, we should raise an exception with the source (self name instance) and an explanation.

**Comment by Abhay-Lejith** [comment]
I am not sure about where I should put this check..Any pointers?

**Comment by fkiraly** [comment]
`BaseTransformer.transform`?
